common/lru/lru_test.go

```diff
@@ -0,0 +1,230 @@
+package lru
+
+import (
+	"container/list"
+	"reflect"
+	"testing"
+)
+
+func (c *Cache) Keys() []interface{} {
+	keys := make([]interface{}, len(c.cache))
+	i := 0
+	for ent := c.ll.Back(); ent != nil; ent = ent.Prev() {
+		keys[i] = ent.Value.(*entry).key
+		i++
+	}
+	return keys
+}
+
+func TestLRU(t *testing.T) {
+	evictCounter := 0
+	onEvicted := func(k Key, v interface{}) {
+		evictCounter++
+	}
+	l := New(128)
+	l.OnEvicted = onEvicted
+	for i := 0; i < 256; i++ {
+		l.Add(i, i, 1)
+	}
+	if l.Len() != 128 {
+		t.Fatalf("bad len: %v", l.Len())
+	}
+
+	if evictCounter != 128 {
+		t.Fatalf("bad evict count: %v", evictCounter)
+	}
+
+	for i, k := range l.Keys() {
+		if v, ok := l.Get(k); !ok || v != k || v != i+128 {
+			t.Fatalf("bad key: %v", k)
+		}
+	}
+	for i := 0; i < 128; i++ {
+		_, ok := l.Get(i)
+		if ok {
+			t.Fatalf("should be evicted")
+		}
+	}
+	for i := 128; i < 256; i++ {
+		_, ok := l.Get(i)
+		if !ok {
+			t.Fatalf("should not be evicted")
+		}
+	}
+	for i := 128; i < 192; i++ {
+		l.Remove(i)
+		l.Remove(i)
+		_, ok := l.Get(i)
+		if ok {
+			t.Fatalf("should be deleted")
+		}
+	}
+
+	l.Get(192) // expect 192 to be last key in l.Keys()
+
+	for i, k := range l.Keys() {
+		if (i < 63 && k != i+193) || (i == 63 && k != 192) {
+			t.Fatalf("out of order key: %v", k)
+		}
+	}
+
+	l.Clear()
+	if l.Len() != 0 {
+		t.Fatalf("bad len: %v", l.Len())
+	}
+	if _, ok := l.Get(200); ok {
+		t.Fatalf("should contain nothing")
+	}
+}
+
+func TestLRU_RemoveOldest(t *testing.T) {
+	l := New(128)
+	for i := 0; i < 256; i++ {
+		l.Add(i, i, 1)
+	}
+}
+
+// Test that Add returns true/false if an eviction occurred
+func TestLRU_Add(t *testing.T) {
+
+	l := New(2)
+	l.Add(1, 1, 1)
+	l.Add(1, 1, 1)
+	if l.Len() > 1 {
+		t.Fatalf("the same cache should not add again")
+	}
+
+	l.Add(1, 1, 1)
+	l.Add(2, 2, 2)
+}
```

Comment:
Add expected result check for unit test, for example 
https://github.com/hashicorp/golang-lru/blob/master/simplelru/lru_test.go#L111
```
// Test that Add returns true/false if an eviction occurred
func TestLRU_Add(t *testing.T) {
	evictCounter := 0
	onEvicted := func(k interface{}, v interface{}) {
		evictCounter++
	}

	l, err := NewLRU(1, onEvicted)
	if err != nil {
		t.Fatalf("err: %v", err)
	}

	if l.Add(1, 1) == true || evictCounter != 0 {
		t.Errorf("should not have an eviction")
	}
	if l.Add(2, 2) == false || evictCounter != 1 {
		t.Errorf("should have an eviction")
	}
}
```

Reply:
If there is no return values for a function, you can check whether the state after function call is expected or not.

Reply:
got it